### ISSUE TITLE
Assignment Service Out of Sync

### ISSUE THREAD

**Issue by ctrimble** [meltmedia/dropwizard-etcd#3]
During a production node restart yesterday, we had an assignment service come out of sync.  This caused processes to stop being migrated between the nodes.

The node that was terminated stopped cleanly and gave up it's processes.  The node controlling assignments picked up some of the workload, but then failed late in the reassignment process.

**Comment by ctrimble** [comment]
During the restart, messages like this can be found in the logs:

```
exception thrown while watching REDACTED! mousio.etcd4j.responses.EtcdException: [401]: The event in requested index is outdated and cleared, cause: the requested history has been cleared [63595677/63595676], at index: 63596676
! at mousio.etcd4j.responses.EtcdKeysResponseParser.parseException(EtcdKeysResponseParser.java:104) ~[talu-twitter-streams-service:na]
! at mousio.etcd4j.responses.EtcdKeysResponseParser.parse(EtcdKeysResponseParser.java:63) ~[talu-twitter-streams-service:na]
! at mousio.etcd4j.transport.EtcdKeyResponseHandler.channelRead0(EtcdKeyResponseHandler.java:61) ~[talu-twitter-streams-service:na]
! at mousio.etcd4j.transport.EtcdKeyResponseHandler.channelRead0(EtcdKeyResponseHandler.java:17) ~[talu-twitter-streams-service:na]
! at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[talu-twitter-streams-service:na]
! at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) ~[talu-twitter-streams-service:na]
! at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153) ~[talu-twitter-streams-service:na]
! at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157) ~[talu-twitter-streams-service:na]
! at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[talu-twitter-streams-service:na]
! at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) ~[talu-twitter-streams-service:na]
! at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153) ~[talu-twitter-streams-service:na]
! at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157) ~[talu-twitter-streams-service:na]
! at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86) ~[talu-twitter-streams-service:na]
! at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) ~[talu-twitter-streams-service:na]
! at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153) ~[talu-twitter-streams-service:na]
! at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157) ~[talu-twitter-streams-service:na]
! at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244) ~[talu-twitter-streams-service:na]
! at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) ~[talu-twitter-streams-service:na]
! at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153) ~[talu-twitter-streams-service:na]
! at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157) ~[talu-twitter-streams-service:na]
! at io.netty.handler.timeout.ReadTimeoutHandler.channelRead(ReadTimeoutHandler.java:150) ~[talu-twitter-streams-service:na]
! at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) ~[talu-twitter-streams-service:na]
! at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153) ~[talu-twitter-streams-service:na]
! at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157) ~[talu-twitter-streams-service:na]
! at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:946) ~[talu-twitter-streams-service:na]
! at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:127) ~[talu-twitter-streams-service:na]
! at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:510) ~[talu-twitter-streams-service:na]
! at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:467) ~[talu-twitter-streams-service:na]
! at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:381) ~[talu-twitter-streams-service:na]
! at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:353) ~[talu-twitter-streams-service:na]
! at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703) ~[talu-twitter-streams-service:na]
! at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) ~[talu-twitter-streams-service:na]
! at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
```

**Comment by ctrimble** [comment]
This message is returned from etcd when you request history that is more than 1000 indices old.  The correct action is to update state and restart the watch.  Not sure why this happened, but perhaps state was arriving faster than we could consume it.

Corrective action:
- Add code to detect condition and restart watches.
- Add metric for watch service.  Would like to see the rate messages arrive in graphite.

**Comment by ctrimble** [comment]
This issue was resolved in version 0.3.0.
